cloudevents/formats/avro-turbo-format.md

```diff
@@ -0,0 +1,217 @@
+
+| CloudEvents   | Avro                                                                   |
+| ------------- | ---------------------------------------------------------------------- |
+| Boolean       | [boolean][avro-primitives]                                             |
+| Integer       | [int][avro-primitives]                                                 |
+| String        | [string][avro-primitives]                                              |
+| Binary        | [bytes][avro-primitives]                                               |
+| URI           | [string][avro-primitives] following [RFC 3986 §4.3][rfc3986-section43] |
+| URI-reference | [string][avro-primitives] following [RFC 3986 §4.1][rfc3986-section41] |
+| Timestamp     | [long][avro-primitives]                                                |
+
+Extension specifications MAY define secondary mapping rules for the values of
+attributes they define, but MUST also include the previously defined primary
+mapping.
+
+### 2.3 OPTIONAL Attributes
+
+CloudEvents Spec defines OPTIONAL attributes. The Avro format defines that these
+fields MUST use the `null` type and the actual type through the
+[union][avro-unions].
+
+Example:
+
+```json
+["null", "string"]
+```
+
+### 2.4 Definition
+
+Users of Avro MUST use a message whose binary encoding is identical to the one
+described by the [CloudEvent Avro Turbo Schema](cloudevents-turbo.avsc):
+
+```json
+{
+  "namespace": "io.cloudevents.v1.avro",
+  "type": "record",
+  "name": "CloudEvent",
+  "version": "1.0",
+  "doc": "Avro Turbo Event Format for CloudEvents",
+  "fields": [
+    {
+      "name": "id",
+      "type": "string"
+    },
+    {
+      "name": "source",
+      "type": "string"
+    },
+    {
+      "name": "type",
+      "type": "string"
+    },
+    {
+      "name": "datacontenttype",
+      "type": [
+        "null",
+        "string"
+      ],
+      "default": null
+    },
+    {
+      "name": "dataschema",
+      "type": [
+        "null",
+        "string"
+      ],
+      "default": null
+    },
+    {
+      "name": "subject",
+      "type": [
+        "null",
+        "string"
+      ],
+      "default": null
+    },
+    {
+      "name": "time",
+      "type": [
+        "null",
+        {
+          "type": "long",
+          "logicalType": "timestamp-millis"
+        }
+      ],
+      "default": null
+    },
+    {
+      "name": "attribute",
+      "type": {
+        "type": "map",
+        "values": [
+          "null",
+          "boolean",
+          "int",
+          {
+            "type": "long",
+            "logicalType": "timestamp-millis"
+          },
+          "string",
+          "bytes"
+        ]
+      },
+      "default": {}
+    },
+    {
+      "name": "data",
+      "type": [
+        "bytes",
+        "null"
+      ],
+      "default": "null"
+    }
+  ]
+}
+```
+
+## 3 Data
+
+Before encoding, the AVRO serializer MUST first determine the runtime data type
```

Comment:
If @JemDay is correct, perhaps just say that the content MUST be serialized as a series of bytes ??
I could see an argument for removing the entire section, but "data" is kind of special so it might be nice to call it out.

Reply:
It looks like this is basically the first half of the equivalent section in the original Avro format.
Both documents probably need more work when we've reached some conclusions on #1204.

(I don't think we can remove the entire section - we've got to say *something* about data. It would be reasonable to say that the payload should be stored in the data field in the same way that it would be encoded to bytes in binary mode, perhaps.)

Reply:
deleting